core/src/transport/upgrade.rs

```diff
@@ -113,6 +113,44 @@ where
             version,
         ))
     }
+
+    /// Upgrades the transport to perform authentication of the remote
+    ///
+    /// The supplied upgrade receives the I/O resource `C` and must
+    /// produce a pair `(PeerId, D)`, where `D` is a new I/O resource.
+    /// The upgrade must thus at a minimum identify the remote, which typically
+    /// involves the use of a cryptographic authentication protocol in the
+    /// context of establishing a secure channel.
+    ///
+    /// ## Transitions
+    ///
+    ///   * I/O upgrade: `C -> (PeerId, D)`.
+    ///   * Transport output: `C -> (PeerId, D)`
+    ///
+    /// ## Alternative state
+    ///
+    /// This function is an alternative code path to [`Builder::authenticate`]
+    /// given the supplied upgrade implements `SecurityUpgrade` instead of
+    /// `InboundConnectionUpgrade`/`OutboundConnectionUpgrade`.
+    pub fn authenticate2<C, D, U, E>(
```

Comment:
Let's use this new function througout the codebase. If we deprecate the old one, we can easily find all usage sites. 

I don't think we want to actually merge a deprecation of `authenticate` though. It might be better to mark this one as `#[doc(hidden)]` and only change the trait-bounds of `authenticate` in the next breaking release.

Reply:
Sure, no problem.